apps/opencs/view/render/instancemode.cpp

```diff
@@ -659,3 +663,24 @@ void CSVRender::InstanceMode::subModeChanged (const std::string& id)
     getWorldspaceWidget().abortDrag();
     getWorldspaceWidget().setSubMode (getSubModeFromId (id), Mask_Reference);
 }
+
+void CSVRender::InstanceMode::deleteSelectedInstances(bool active)
+{
+    std::vector<osg::ref_ptr<TagBase> > selection = getWorldspaceWidget().getSelection (Mask_Reference);
+    if (selection.empty()) return;
+
+    CSMDoc::Document& document = getWorldspaceWidget().getDocument();
+    CSMWorld::IdTable& referencesTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_References));
+    QUndoStack& undoStack = document.getUndoStack();
+
+    undoStack.beginMacro ("Delete Instances");
```

Comment:
This should be using CSMWorld::CommandMacro instead. Things like begin/end, open/close, lock/unlock are an anti-pattern in C++.

Reply:
Looks like CSMWorld::CommandMacro is the same thing, but a bit stricter when it comes to hanging leaving macro hanging open. Good advice! This rings a bell though, I've used beginMacro directly in terrainshapemode, and possible in terraintexturemode, so it's probably a good idea to update them to use CSMWorld::CommandMacro too. Great to see you back!